### DISCUSSION TITLE
JSONApi nested attributes

### DISCUSSION THREAD

**Discussion by eedijs** [yahoo/elide#2810] [Q&A]
Hi,

Is there a plan to add the possibility to have nested attributes? At the moment this `Attributes` class can only be created from an array of `Attribute` objects, but what if there was another constructor that accepted another `Attributes` object or similar, that also contains an array of `Attribute` objects?

```java
public class Attributes extends LinkedHashMap<String, Object> {

    /**
     * Instantiates a new Attributes.
     *
     * @param attributes the attributes
     */
    public Attributes(Attribute... attributes) {
        for (Attribute attribute : attributes) {
            this.put(attribute.key, attribute.value);
        }
    }
}
```

Example usage:

```java
        datum(
                resource(
                        type("something"),
                        attributes(
                                attr("attribute", "value"),
                                attributes(
                                        attr("additionalAttribute", "additionalValue")
                                )
                        )
                )
        )
```

**Comment by jkusa** [comment]
Hi @eedijs 👋  Are you asking about nested attributes in terms of testing or in general? 

In general you can use a custom type in hibernate to serialize/deserialize to/from a string/fragment class.

Here is an example of a Person entity with an address fragment: 
- [Person.java](https://github.com/yahoo/elide/blob/113e7f2e821fb83eeb2915e61c42b81ee659db67/elide-datastore/elide-datastore-jpa/src/test/java/example/Person.java)
- [AddressFragment.java](https://github.com/yahoo/elide/blob/113e7f2e821fb83eeb2915e61c42b81ee659db67/elide-datastore/elide-datastore-jpa/src/test/java/example/AddressFragment.java)
- [JsonType.java](https://github.com/yahoo/elide/blob/113e7f2e821fb83eeb2915e61c42b81ee659db67/elide-datastore/elide-datastore-jpa/src/test/java/com/yahoo/elide/datastores/jpa/usertypes/JsonType.java)

For testing with the test helpers you can provide an instance of the fragment class as seen here (in Kotlin):

```kt
        given()
            .header("User", USER1)
            .contentType("application/vnd.api+json")
            .body(
                datum(
                    resource(
                        type("deliveryRules"),
                        id("1"),
                        attributes(
                            attr("frequency", "quarter"),
                            attr("schedulingRules", SchedulingRules(true)), // sets attribute with new instance of SchedulingRules fragment class
                        )
                    )
                ).toJSON()
            )
            .`when`()
            .patch("/deliveryRules/1")
            .then()
            .assertThat()
            .statusCode(HttpStatus.SC_NO_CONTENT)
```

[DeliveryRulesTest.kt](https://github.com/yavin-dev/framework/blob/a439969799af8f4ed81e73ba3a135328bdb85b99/packages/webservice/app/src/test/kotlin/com/yahoo/navi/ws/test/integration/DeliveryRulesTest.kt#L301-L320)

**Reply by eedijs** [reply]
Hi,

Yes, this is for unit testing, but with a real class that is used in production.

I didn't know that you could even pass a whole instance in attributes, just tried that it might even be possible to also just put a `Map.of(something)` in there.

Thanks, I think this solves the issue.
